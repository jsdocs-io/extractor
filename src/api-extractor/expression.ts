import * as tsm from 'ts-morph';
import {
    DeclarationKinds,
    VariableDeclaration,
} from '../types/module-declarations';
import { getApparentType } from './get-apparent-type';
import { getJSDocs } from './get-jsdocs';
import { SourceProvider } from './source-provider';
import { getVariableSignature, newVariable } from './variables';

export function isVariableAssignmentExpression(
    declaration: tsm.Node
): declaration is tsm.BinaryExpression {
    return (
        tsm.Node.isBinaryExpression(declaration) &&
        tsm.Node.isIdentifier(declaration.getLeft())
    );
}

export function newVariableAssignmentExpression({
    id,
    name,
    declaration,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.BinaryExpression;
    getSource: SourceProvider;
}): VariableDeclaration {
    const variableDeclaration = getVariableDeclaration({ declaration });
    const suggestedType = getApparentType({
        declaration: declaration.getRight(),
    });

    return newVariable({
        id,
        name,
        declaration: variableDeclaration,
        getSource,
        suggestedType,
    });
}

function getVariableDeclaration({
    declaration,
}: {
    declaration: tsm.BinaryExpression;
}): tsm.VariableDeclaration {
    return declaration
        .getLeft()
        .getSymbol()!
        .getDeclarations()[0] as tsm.VariableDeclaration;
}

export function isExpression(
    declaration: tsm.Node
): declaration is tsm.Expression {
    return tsm.Node.isExpression(declaration);
}

export function newExpression({
    id,
    name,
    declaration,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.Expression;
    getSource: SourceProvider;
}): VariableDeclaration {
    const kind = DeclarationKinds.VariableDeclaration;
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const variableKind = 'const';
    const type = getApparentType({ declaration });
    const signature = getVariableSignature({ variableKind, name, type });

    return {
        kind,
        id,
        name,
        docs,
        source,
        signature,
        variableKind,
        type,
    };
}
