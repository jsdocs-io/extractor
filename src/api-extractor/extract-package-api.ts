import { performance } from "perf_hooks";
import * as tsm from "ts-morph";
import { PackageAPI } from "../types/package-api";
import { TaggedGitRepository } from "../types/tagged-git-repository";
import { log } from "../utils/log";
import { getOverview } from "./get-overview";
import { getPackageDeclarations } from "./get-package-declarations";
import { getPackageFiles } from "./get-package-files";
import { getProject } from "./get-project";
import { getRepositoryFileURLProvider } from "./repository-file-url-provider";
import { getSourceProvider } from "./source-provider";
import { getTypeChecker } from "./type-checker";
import { getUnpkgFileURLProvider } from "./unpkg-file-url-provider";

/**
 * `extractPackageAPI` extracts the public API from a package.
 *
 * @param fileSystem - filesystem containing the package's source code
 * @param entryPoint - absolute path of the file acting as the package's entry point
 * @param maxDepth - maximum depth for analyzing nested namespaces (default: `5`)
 * @param pattern - file pattern including files to be analyzed
 * @param repository - a tagged git repository to enable linking to source
 * @param id - npm-style package ID used for logging (for example, `foo@1.0.0`)
 *
 * @see {@link PackageAPI}
 */
export function extractPackageAPI({
  fileSystem,
  entryPoint,
  maxDepth,
  pattern,
  repository,
  id,
}: {
  fileSystem: tsm.FileSystemHost;
  entryPoint: string;
  maxDepth?: number;
  pattern?: string;
  repository?: TaggedGitRepository;
  id?: string;
}): PackageAPI {
  const start = performance.now();
  log("extractPackageAPI: extracting API: %O", {
    id,
    entryPoint,
    maxDepth,
    pattern,
    repository,
    fileSystem,
  });

  const project = getProject({ fileSystem, pattern });
  log("extractPackageAPI: created project: %O", {
    id,
    files: project.getSourceFiles().map((file) => file.getFilePath()),
  });

  const indexFile = project.getSourceFileOrThrow(entryPoint);
  log("extractPackageAPI: found index file: %O", {
    id,
    indexFile: indexFile.getFilePath(),
  });

  const getRepositoryFileURL = getRepositoryFileURLProvider({ repository });
  log("extractPackageAPI: got repository file URL provider");

  const getUnpkgFileURL = getUnpkgFileURLProvider({ id });
  log("extractPackageAPI: got unpkg file URL provider");

  const getSource = getSourceProvider({
    getRepositoryFileURL,
    getUnpkgFileURL,
  });
  log("extractPackageAPI: got source provider");

  const getType = getTypeChecker({ project });
  log("extractPackageAPI: got type checker");

  const overview = getOverview({ indexFile });
  log("extractPackageAPI: got package overview: %O", { id, overview });

  const declarations = getPackageDeclarations({
    project,
    indexFile,
    maxDepth,
    getSource,
    getType,
  });
  log("extractPackageAPI: got package declarations: %O", {
    id,
    declarations,
  });

  const files = getPackageFiles({
    indexFile,
    declarations,
    getRepositoryFileURL,
    getUnpkgFileURL,
  });
  log("extractPackageAPI: got package files: %O", { id, files });

  const elapsed = Math.round(performance.now() - start);
  log("extractPackageAPI: performance (ms): %O", { id, elapsed });

  log("extractPackageAPI: extracted package API: %O", {
    id,
    api: { overview, declarations, files },
  });

  return { overview, declarations, files };
}
