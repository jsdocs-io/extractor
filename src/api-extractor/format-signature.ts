import { format } from "prettier";
import type { DeclarationKind } from "../types/declaration-kind";

export const formatSignature = async (
  kind: DeclarationKind,
  signature: string
): Promise<string> => {
  const s = signature
    .trim()
    // Remove unnecessary keywords at declaration start.
    .replace(/^export\s+/, "")
    .replace(/^default\s+/, "")
    .replace(/^declare\s+/, "");
  switch (kind) {
    case "variable": {
      // Temporarily replace the invalid variable name `default` before the type.
      const raw = s.replace("default:", "_______:");
      const formatted = await formatWithPrettier(raw);
      return formatted.replace("_______:", "default:");
    }
    case "function": {
      // Temporarily add `let` to format the function like a variable signature.
      const raw = `let ${s.replace("default:", "_______:")}`;
      const formatted = await formatWithPrettier(raw);
      return formatted.replace("_______:", "default:").replace(/^let\s/, "");
    }
    case "class":
    case "interface":
    case "enum":
    case "typeAlias":
    case "namespace": {
      // Just format the signature as-is.
      const formatted = await formatWithPrettier(s);
      return formatted;
    }
    case "classConstructor":
    case "classProperty":
    case "classMethod":
    case "interfaceProperty":
    case "interfaceMethod":
    case "interfaceConstructSignature":
    case "interfaceCallSignature":
    case "interfaceIndexSignature":
    case "enumMember": {
      // Temporarily wrap members in their parent declaration for formatting.
      const parentKind = kind.split(/[A-Z]/)[0]!;
      const raw = `${parentKind} P { ${s} }`;
      const formattedParent = await formatWithPrettier(raw);
      const formatted = formattedParent
        // Remove parent wrapper on first and last lines.
        .split("\n")
        .slice(1, -1)
        // Remove indentation.
        .map((line) => line.replace(/^\s{4}/, ""))
        .join("\n")
        // Remove trailing comma in enum member signature.
        .replace(/,$/, "");
      return formatted;
    }
  }
};

const formatWithPrettier = async (s: string): Promise<string> => {
  try {
    return (
      await format(s, {
        semi: true,
        singleQuote: true,
        trailingComma: "es5",
        tabWidth: 4,
        printWidth: 85,
        endOfLine: "lf",
        arrowParens: "always",
        parser: "typescript",
      })
    ).trim();
  } catch {
    return s;
  }
};
