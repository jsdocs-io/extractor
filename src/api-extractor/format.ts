import * as prettier from "prettier";
import { log } from "../utils/log";

const formatOptions: prettier.Options = {
  semi: true,
  singleQuote: true,
  trailingComma: "es5",
  tabWidth: 4,
  printWidth: 85,
  endOfLine: "lf",
  arrowParens: "always",
  parser: "typescript",
};

export function formatFunctionSignature(text: string): string {
  const varLike = `let ${text}`;
  return formatVariableSignature(varLike).replace(/^let\s/, "");
}

export function formatVariableSignature(text: string): string {
  // Temporarily replace the invalid variable name `default`
  const escapedText = text.replace("default:", "_______:");
  return formatText(escapedText).replace("_______:", "default:");
}

export function formatClassMember(text: string): string {
  const wrapped = `class C { ${text} }`;
  return formatWrappedMember(wrapped);
}

export function formatInterfaceMember(text: string): string {
  const wrapped = `interface I { ${text} }`;
  return formatWrappedMember(wrapped);
}

export function formatEnumMember(text: string): string {
  const wrapped = `enum E { ${text} }`;
  return formatWrappedMember(wrapped).replace(/,$/, "");
}

function formatWrappedMember(text: string): string {
  const formatted = formatText(text);
  const lines = formatted.split("\n");

  // Remove wrapper and member indentation
  const member = lines
    .slice(1, lines.length - 1)
    .map((line) => line.replace(/^\s{4}/, ""))
    .join("\n");

  return member;
}

export function formatText(text: string): string {
  let formatted = text
    .trim()
    .replace(/^export\s+/, "")
    .replace(/^default\s+/, "")
    .replace(/^declare\s+/, "");

  try {
    formatted = prettier.format(formatted, formatOptions);
  } catch (err) {
    // istanbul ignore next
    log("formatText: formatting error: %O", { err });
  }

  return formatted.trim();
}
