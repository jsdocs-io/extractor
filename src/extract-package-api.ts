import { Effect } from "effect";
import type { ExtractedDeclaration } from "./extract-declarations";
import { extractPackageApiEffect } from "./extract-package-api-effect";

/**
`ExtractPackageApiOptions` contains all the options
for calling {@link extractPackageApi}.
*/
export type ExtractPackageApiOptions = {
	/**
  Package to extract the API from.

  This can be either a package name (e.g., `foo`) or any other query
  that can be passed to `bun add` (e.g., `foo@1.0.0`).

  @see {@link https://bun.sh/docs/cli/add | Bun docs}
  */
	pkg: string;

	/**
  Specific subpath to consider in a package.

  If a package has multiple entrypoints listed in the `exports` property
  of its `package.json`, use `subpath` to select a specific one by its name
  (e.g., `someFeature`).

  @defaultValue `.` (package root)

  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  @see {@link https://github.com/lukeed/resolve.exports | resolve.exports docs}
  */
	subpath?: string;

	/**
  Packages can have deeply nested modules and namespaces.

  Use `maxDepth` to limit the depth of the extraction.
  Declarations nested at levels deeper than this value will be ignored.

  @defaultValue 5
  */
	maxDepth?: number;

	/**
  Absolute path to the `bun` executable.
  Used to locate bun if it's not in `PATH`.

  @defaultValue `bun`
  */
	bunPath?: string;
};

/**
`PackageApi` contains all the information extracted from a package.
*/
export type PackageApi = {
	/** Package name (e.g., `foo`). */
	name: string;

	/** Package version number (e.g., `1.0.0`). */
	version: string;

	/**
  Package subpath selected when extracting the API (e.g., `.`, `someFeature`).

  @see {@link ExtractPackageApiOptions.subpath}
  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  */
	subpath: string;

	/**
  Type declarations file, resolved from the selected `subpath`,
  that acts as the entrypoint for the package (e.g., `index.d.ts`).
  */
	types: string;

	/**
  Package description extracted from the `types` file if a
  JSDoc comment with the `@packageDocumentation` tag is found.
  */
	overview: string | undefined;

	/** Declarations exported (or re-exported) by the package. */
	declarations: ExtractedDeclaration[];

	/**
  All packages resolved and installed when installing the package (included).

  @example
  ```ts
  ["foo@1.0.0", "bar@2.0.0", "baz@3.0.0"]
  ```
  */
	packages: string[];

	/** Timestamp of when the package was analyzed. */
	analyzedAt: string;

	/** Package analysis duration in milliseconds. */
	analyzedIn: number;
};

/**
`extractPackageApi` extracts the API from a package.

If the extraction succeeds, `extractPackageApi` returns a {@link PackageApi} object.
If the extraction fails, `extractPackageApi` throws an error.

Warning: The extraction process is slow and blocks the main thread, using workers is recommended.

@example
```ts
const packageApi = await extractPackageApi({
  pkg: "foo",    // Extract API from npm package `foo` [Required]
  subpath: ".",  // Select subpath `.` (root subpath) [Optional]
  maxDepth: 5,   // Maximum depth for analyzing nested namespaces [Optional]
  bunPath: "bun" // Absolute path to the `bun` executable [Optional]
});
console.log(JSON.stringify(packageApi, null, 2));
```

@param options - {@link ExtractPackageApiOptions}

@returns A {@link PackageApi} object
*/
export const extractPackageApi = ({
	pkg,
	subpath = ".",
	maxDepth = 5,
	bunPath = "bun",
}: ExtractPackageApiOptions): Promise<PackageApi> =>
	Effect.runPromise(
		extractPackageApiEffect({ pkg, subpath, maxDepth, bunPath }),
	);
