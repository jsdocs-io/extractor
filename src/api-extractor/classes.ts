import * as tsm from 'ts-morph';
import {
    isClassMethodDeclaration,
    isClassPropertyDeclaration,
} from '../types/declaration-type-guards';
import {
    ClassConstructorDeclaration,
    ClassDeclaration,
    ClassMemberDeclarations,
    ClassMethodDeclaration,
    ClassPropertyDeclaration,
} from '../types/module-declarations';
import { formatClassMember } from './format';
import { getApparentType } from './get-apparent-type';
import { getJSDocs } from './get-jsdocs';
import { getModifiersText } from './get-modifiers-text';
import { getWrapperSignature } from './get-wrapper-signature';
import { isInternalDeclaration } from './is-internal-declaration';
import { sortByID } from './sort-by-id';
import { SourceProvider } from './source-provider';
import { toID } from './to-id';
import { TypeChecker } from './type-checker';

export function isClass(
    declaration: tsm.Node
): declaration is tsm.ClassDeclaration {
    return tsm.Node.isClassDeclaration(declaration);
}

export function newClass({
    id,
    name,
    declaration,
    getSource,
    getType,
}: {
    id: string;
    name: string;
    declaration: tsm.ClassDeclaration;
    getSource: SourceProvider;
    getType: TypeChecker;
}): ClassDeclaration {
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const isAbstract = declaration.isAbstract();
    const signature = getWrapperSignature({ declaration });

    const constructors = getClassConstructors({
        classID: id,
        classDeclaration: declaration,
        getSource,
    });

    const members = getClassMembers({
        classID: id,
        classDeclaration: declaration,
        getSource,
        getType,
    });

    return {
        kind: 'class',
        id,
        name,
        docs,
        source,
        signature,
        isAbstract,
        constructors,
        members,
    };
}

function getClassConstructors({
    classID,
    classDeclaration,
    getSource,
}: {
    classID: string;
    classDeclaration: tsm.ClassDeclaration;
    getSource: SourceProvider;
}): ClassConstructorDeclaration[] {
    // `getConstructors()` returns all constructors for ambient modules
    // but only the implementation constructor in normal modules
    const declaration = classDeclaration.getConstructors()[0];
    if (!declaration) {
        return [];
    }

    // Manually retrieve all constructors from the first constructor
    const overloads = declaration.getOverloads();
    const implementation = declaration.getImplementation();
    const constructors = [
        ...overloads,
        ...(implementation ? [implementation] : []),
    ];

    return constructors.flatMap((declaration, index) => {
        if (isInternalDeclaration({ declaration })) {
            return [];
        }

        const name = 'constructor';
        const id = toID(classID, `${index}-${name}`);
        const docs = getJSDocs({ declaration });
        const source = getSource({ declaration });
        const signature = getClassConstructorSignature({ declaration });

        return {
            kind: 'classConstructor',
            id,
            name,
            docs,
            source,
            signature,
        };
    });
}

function getClassConstructorSignature({
    declaration,
}: {
    declaration: tsm.ConstructorDeclaration;
}): string {
    const modifiers = declaration
        .getModifiers()
        .map((modifier) => modifier.getText())
        .join(' ');

    const params = declaration
        .getParameters()
        .map((param) => {
            const name = param.getName();
            const type = getApparentType({ declaration: param });
            const isRest = param.isRestParameter();
            const dotsToken = isRest ? '...' : '';
            const isOptional = param.isOptional();
            const questionToken = !isRest && isOptional ? '?' : '';
            return `${dotsToken}${name}${questionToken}: ${type}`;
        })
        .join(',');

    const signature = `${modifiers} constructor(${params});`;
    return formatClassMember(signature);
}

function getClassMembers({
    classID,
    classDeclaration,
    getSource,
    getType,
}: {
    classID: string;
    classDeclaration: tsm.ClassDeclaration;
    getSource: SourceProvider;
    getType: TypeChecker;
}): ClassMemberDeclarations {
    const seenMethods = new Set<string>();
    const members = [
        ...classDeclaration.getStaticMembers(),
        ...classDeclaration.getInstanceMembers(),
    ]
        .flatMap((declaration) => {
            const name = declaration.getName();
            const id = toID(classID, name);

            if (isInternalDeclaration({ declaration, name })) {
                return [];
            }

            if (
                tsm.Node.isPropertyDeclaration(declaration) ||
                tsm.Node.isParameterDeclaration(declaration)
            ) {
                return newProperty({ id, name, declaration, getSource });
            }

            if (tsm.Node.isGetAccessorDeclaration(declaration)) {
                return newGetAccessor({ id, name, declaration, getSource });
            }

            if (tsm.Node.isMethodDeclaration(declaration)) {
                // Skip overloaded methods
                if (seenMethods.has(id)) {
                    return [];
                }

                seenMethods.add(id);
                return newMethod({ id, name, declaration, getSource, getType });
            }

            return [];
        })
        .sort(sortByID);

    return {
        properties: members.filter(isClassPropertyDeclaration),
        methods: members.filter(isClassMethodDeclaration),
    };
}

function newProperty({
    id,
    name,
    declaration,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.PropertyDeclaration | tsm.ParameterDeclaration;
    getSource: SourceProvider;
}): ClassPropertyDeclaration {
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const modifiersText = getModifiersText({ declaration });
    const isStatic =
        tsm.Node.isPropertyDeclaration(declaration) && declaration.isStatic();
    const isOptional = declaration.hasQuestionToken();
    const optionalText = isOptional ? '?' : '';
    const type = getApparentType({ declaration });
    const signature = formatClassMember(
        `${modifiersText} ${name} ${optionalText}: ${type}`
    );

    return {
        kind: 'classProperty',
        id,
        name,
        docs,
        source,
        signature,
        isStatic,
        type,
    };
}

function newGetAccessor({
    id,
    name,
    declaration,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.GetAccessorDeclaration;
    getSource: SourceProvider;
}): ClassPropertyDeclaration {
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const isStatic = declaration.isStatic();
    const isReadonly = declaration.getSetAccessor() === undefined;
    const type = getApparentType({ declaration: declaration });
    const staticText = isStatic ? 'static' : '';
    const readonlyText = isReadonly ? 'readonly' : '';
    const signature = formatClassMember(
        `${staticText} ${readonlyText} ${name}: ${type}`
    );

    return {
        kind: 'classProperty',
        id,
        name,
        docs,
        source,
        signature,
        isStatic,
        type,
    };
}

function newMethod({
    id,
    name,
    declaration,
    getSource,
    getType,
}: {
    id: string;
    name: string;
    declaration: tsm.MethodDeclaration;
    getSource: SourceProvider;
    getType: TypeChecker;
}): ClassMethodDeclaration {
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const isStatic = declaration.isStatic();
    const modifiersText = getModifiersText({ declaration });
    const type = getType({ declaration });
    const signature = formatClassMember(`${modifiersText} ${name}: ${type}`);

    return {
        kind: 'classMethod',
        id,
        name,
        docs,
        source,
        signature,
        isStatic,
        type,
    };
}
