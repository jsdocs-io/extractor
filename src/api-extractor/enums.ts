import * as tsm from 'ts-morph';
import {
    Declaration,
    DeclarationKinds,
    EnumDeclaration,
    EnumMemberDeclaration,
} from '../types/module-declarations';
import { formatEnumMember, formatText } from './format';
import { getJSDocs } from './get-jsdocs';
import { isInternalDeclaration } from './is-internal-declaration';
import { sortByID } from './sort-by-id';
import { SourceProvider } from './source-provider';
import { toID } from './to-id';

export function isEnum(
    declaration: tsm.Node
): declaration is tsm.EnumDeclaration {
    return tsm.Node.isEnumDeclaration(declaration);
}

export function newEnum({
    id,
    name,
    declaration,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.EnumDeclaration;
    getSource: SourceProvider;
}): EnumDeclaration {
    const kind = DeclarationKinds.EnumDeclaration;
    const docs = getJSDocs({ declaration });
    const source = getSource({ declaration });
    const isConst = declaration.isConstEnum();

    // Keep members in original order for signature text and sort them later
    const members = getEnumMembers({
        enumID: id,
        enumDeclaration: declaration,
        getSource,
    });
    const signature = getEnumSignature({ isConst, name, members });
    members.sort(sortByID);

    return {
        kind,
        id,
        name,
        docs,
        source,
        signature,
        isConst,
        members,
    };
}

function getEnumMembers({
    enumID,
    enumDeclaration,
    getSource,
}: {
    enumID: string;
    enumDeclaration: tsm.EnumDeclaration;
    getSource: SourceProvider;
}): EnumMemberDeclaration[] {
    return enumDeclaration.getMembers().flatMap((declaration) => {
        const name = declaration.getName();
        if (isInternalDeclaration({ declaration, name })) {
            return [];
        }

        const kind = DeclarationKinds.EnumMemberDeclaration;
        const id = toID(enumID, name);
        const docs = getJSDocs({ declaration });
        const source = getSource({ declaration });
        const signature = formatEnumMember(declaration.getText());
        const value = declaration.getValue();

        return {
            kind,
            id,
            name,
            docs,
            source,
            signature,
            value,
        };
    });
}

function getEnumSignature({
    isConst,
    name,
    members,
}: {
    isConst: boolean;
    name: string;
    members: Declaration[];
}): string {
    const kind = isConst ? 'const enum' : 'enum';
    const membersText = members.map(({ signature }) => signature).join();
    const signature = `${kind} ${name} { ${membersText} }`;
    return formatText(signature);
}
