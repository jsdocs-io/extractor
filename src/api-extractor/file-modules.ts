import * as tsm from 'ts-morph';
import {
    DeclarationKinds,
    ModuleDeclarations,
    NamespaceDeclaration,
} from '../types/module-declarations';
import { formatText } from './format';
import { getFilename } from './get-filename';
import { SourceProvider } from './source-provider';

export function isFileModule(
    declaration: tsm.Node
): declaration is tsm.SourceFile {
    return tsm.Node.isSourceFile(declaration);
}

export function newFileModule({
    id,
    name,
    declaration,
    declarations,
    getSource,
}: {
    id: string;
    name: string;
    declaration: tsm.SourceFile;
    declarations: ModuleDeclarations;
    getSource: SourceProvider;
}): NamespaceDeclaration {
    const kind = DeclarationKinds.NamespaceDeclaration;
    const docs = getFileModuleDocs({ declaration });
    const source = getSource({ declaration });
    const signature = getFileModuleSignature({ declaration });

    return {
        kind,
        id,
        name,
        docs,
        source,
        signature,
        declarations,
    };
}

function getFileModuleDocs({
    declaration,
}: {
    declaration: tsm.SourceFile;
}): string[] {
    const doc = declaration
        .getDescendantsOfKind(tsm.SyntaxKind.JSDoc)[0]
        ?.getText();

    return doc ? [doc] : [];
}

function getFileModuleSignature({
    declaration,
}: {
    declaration: tsm.SourceFile;
}): string {
    const filename = getFilename({ declaration });
    const signature = `module '${filename}' {}`;
    return formatText(signature);
}
