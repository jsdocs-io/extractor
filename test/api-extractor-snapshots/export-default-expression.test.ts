import { beforeAll, describe, expect, it } from "vitest";
import { extractPackageAPI } from "../../src/api-extractor/extract-package-api";
import type { PackageAPI } from "../../src/types/package-api";

import { getTestFileSystem } from "../helpers/get-test-file-system";

describe("export-default-expression", () => {
  let api: PackageAPI;

  beforeAll(async () => {
    const name = "export-default-expression";
    const fileSystem = getTestFileSystem({ name });
    api = await extractPackageAPI({ fileSystem, entryPoint: "index.ts" });
  });

  it("snapshot", () => {
    expect(api).toMatchSnapshot();
  });
});
