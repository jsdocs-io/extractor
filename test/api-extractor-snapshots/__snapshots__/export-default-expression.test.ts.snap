// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export-default-expression > snapshot 1`] = `
{
  "declarations": {
    "classes": [],
    "enums": [],
    "functions": [],
    "interfaces": [],
    "namespaces": [],
    "typeAliases": [],
    "variables": [
      {
        "docs": [
          "/** expr1 */",
        ],
        "id": "default",
        "kind": "VariableDeclaration",
        "name": "default",
        "signature": "const default: any;",
        "source": {
          "filename": "index.ts",
          "line": 2,
          "unpkgURL": undefined,
          "url": undefined,
        },
        "type": "any",
        "variableKind": "const",
      },
    ],
  },
  "files": [
    {
      "filename": "index.ts",
      "isIndexFile": true,
      "unpkgURL": undefined,
      "url": undefined,
    },
  ],
  "overview": undefined,
}
`;
